codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -167,11 +167,29 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                 }
             }
 
-            // Pass headers to the flow-controller so it can maintain their sequence relative to DATA frames.
-            flowController().addFlowControlled(stream,
-                    new FlowControlledHeaders(stream, headers, streamDependency, weight, exclusive, padding,
-                            endOfStream, promise));
-            return promise;
+            // Trailing headers must go through flow control if there are other frames queued in flow control
+            // for this stream.
+            Http2RemoteFlowController flowController = flowController();
+            if (!endOfStream || !flowController.hasFlowControlled(stream)) {
```

Comment:
From the HTTP/2 perspective IDLE streams are those whose initial headers frame needs to be ordered. That [seems to](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L925) correspond to the [`createStream()` call earlier in this method](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L154). Keying off of that seems highly-reliable and doesn't require any additional state management.

I'll note that there can be headers other than the initial headers for the stream and trailers, even when only considering HTTP/1.1 compatible because of the 1xx status codes. Not worth special-casing, but technically they should work.
https://httpwg.github.io/specs/rfc7540.html#rfc.section.8.1